keps/sig-storage/3141-prevent-volume-mode-conversion/README.md

```diff
@@ -0,0 +1,820 @@
+
+### Changes to external-provisioner
+
+This design leverages the access rights of a user on `VolumeSnapshotContents` to
+determine whether the volume mode can be modified during a restore operation.
+The volume mode can be altered if the requesting user has rights to `Update` and
+`Patch` a`VolumeSnapshotContent` (which is a cluster scoped resource).
+The control flow of a restore operation for a `PVC` from a `VolumeSnapshot` will 
+look like below:
+
+1. A user attempts to restore a `PVC` from a `VolumeSnapshot` by specifying the
+`Spec.DataSource` parameter of the `PVC` YAML.
+2. The `external-provisioner` receives a callback to dynamically create the volume.
+As part of the preprocessing steps, it will:
+   1. Get the `Spec.SourceVolumeMode` of the `VolumeSnapshotContent`.
+      1. If `Spec.SourceVolumeMode` doesn't exist or is nil, then continue with
+      volume provisioning to preserve existing behavior.
+   2. Get the `Spec.VolumeMode` of the `PVC` being restored.
+   If they do not match:
+      1. Get all annotations on the `VolumeSnapshotContent` and verify if 
+      `snapshot.storage.kubernetes.io/allowVolumeModeChange: true` exists.
+      If it does not exist, block volume provisioning by returning an error.
+4. In all other cases, let volume provisioning continue.
+
+NOTE: `external-provisioner` maintains a reference to `PVC` and `VolumeSnapshotContent`
+during volume creation. This proposal leverages those references to make additional
+decisions.
+
+### Test Plan
+
+E2E tests will be added for this design, that attempt to restore a volume with
+and without requisite privileges. 
+
+### Graduation Criteria
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Define graduation milestones.
+
+These may be defined in terms of API maturity, or as something else. The KEP
+should keep this high-level with a focus on what signals will be looked at to
+determine graduation.
+
+Consider the following in developing the graduation criteria for this enhancement:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc
+definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
+or by redefining what graduation means.
+
+In general we try to use the same stages (alpha, beta, GA), regardless of how the
+functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
```

Comment:
What are features A, B, C?

Reply:
This was actually a comment from the KEP template. I've removed it to avoid confusion.